pkg/estimator/client/service.go

```diff
@@ -0,0 +1,68 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+)
+
+// ResolveCluster parses Service resource content by itself.
+// Fixes Issue https://github.com/karmada-io/karmada/issues/2487
+// Modified from "k8s.io/apiserver/pkg/util/proxy/proxy.go:92 => func ResolveCluster"
+func resolveCluster(kubeClient kubernetes.Interface, namespace, id string, port int32) (string, error) {
+	svc, err := kubeClient.CoreV1().Services(namespace).Get(context.TODO(), id, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			/*
+			 * When Deploying Karmada in Host Kubernetes Cluster, the kubeClient will connect kube-apiserver
+			 * of Karmada Control Plane, rather than of host cluster.
+			 * But the Service resource is defined in Host Kubernetes Cluster. So we cannot get its content here.
+			 * The best thing we can do is just glue host:port together, and try to connect to it.
+			 */
+			return net.JoinHostPort(id, fmt.Sprintf("%d", port)), nil
+		}
+
+		return "", err
+	}
+
+	switch {
+	case svc.Spec.Type == v1.ServiceTypeClusterIP && svc.Spec.ClusterIP == v1.ClusterIPNone:
+		return "", fmt.Errorf(`cannot route to service with ClusterIP "None"`)
+	// use IP from a clusterIP for these service types
+	case svc.Spec.Type == v1.ServiceTypeClusterIP, svc.Spec.Type == v1.ServiceTypeLoadBalancer,
```

Comment:
> BTW, if so, I think only external service will take effect. A clusterIP or nodePort service created in karmada control plane will never be connected cause the IP is fake.

Got it, I mean here, the svc is created in karmada-apiserver, so the cluster IP and node port is fake, right? We don't have any kube-proxy in karmada control plane.

Reply:
I copied the logic from "k8s.io/apiserver/pkg/util/proxy/proxy.go:92 => func ResolveCluster", and didn't delete this logic.

I think ClusterIP, LoadBalancer, NodePort will not be used.

Maybe I should delete them?

Reply:
Yes, these types will never be used. I think we could delete them.